google/cloud/ndb/_cache.py

```diff
@@ -318,9 +318,36 @@ def add(self, key, value):
         """
         future = tasklets.Future(info=self.future_info(key, value))
         self.todo[key] = value
-        self.futures.append(future)
+        self.futures[key] = future
```

Comment:
This can result in an unfinished future if multiple of the same key are in a batch.